plonky2/src/hash/field_merkle_tree.rs

```diff
@@ -0,0 +1,322 @@
+    /// The data stored in the Merkle tree leaves.
+    pub leaves: Vec<Vec<Vec<F>>>,
+
+    /// Merkle tree node hashes, analogous to `digests` in `MerkleTree`.
+    pub digests: Vec<H::Hash>,
+
+    /// Represents the roots of the Merkle tree. This allows for using any layer as the root of the tree.
+    pub cap: MerkleCap<F, H>,
+
+    /// Represents the heights at which leaves reside within the tree.
+    pub leaf_heights: Vec<usize>,
+}
+
+impl<F: RichField, H: Hasher<F>> Default for FieldMerkleTree<F, H> {
+    fn default() -> Self {
+        Self {
+            leaves: Vec::new(),
+            digests: Vec::new(),
+            cap: MerkleCap::default(),
+            leaf_heights: Vec::new(),
+        }
+    }
+}
+
+impl<F: RichField, H: Hasher<F>> FieldMerkleTree<F, H> {
+    /// Each element in the `leaves` vector represents a matrix (a vector of vectors).
+    /// The height of each matrix should be a power of two.
+    /// The `leaves` vector should be sorted by matrix height, from tallest to shortest, with no duplicate heights.
+    pub fn new(mut leaves: Vec<Vec<Vec<F>>>, cap_height: usize) -> Self {
+        assert!(!leaves.is_empty());
+        assert!(leaves.iter().all(|leaf| leaf.len().is_power_of_two()));
+        assert!(leaves
+            .windows(2)
+            .all(|pair| { pair[0].len() > pair[1].len() }));
+
+        let leaves_len = leaves[0].len();
+        let log2_leaves_len = log2_strict(leaves_len);
+        assert!(
+            cap_height <= log2_leaves_len,
+            "cap_height={} should be at most log2(leaves.len())={}",
+            cap_height,
+            log2_leaves_len
+        );
+
+        let mut leaf_heights = vec![];
+
+        let num_digests = 2 * (leaves_len - (1 << cap_height));
+        let mut digests = Vec::with_capacity(num_digests);
+        let digests_buf = capacity_up_to_mut(&mut digests, num_digests);
+        let mut digests_buf_pos = 0;
+
+        let mut cap = vec![];
+        let dummy_leaves = vec![vec![F::ZERO]; 1 << cap_height];
+        leaves.push(dummy_leaves);
+        for window in leaves.windows(2) {
+            let cur = &window[0];
+            let next = &window[1];
+
+            let cur_leaf_len = cur.len();
+            let next_cap_len = next.len();
+            let next_cap_height = log2_strict(next_cap_len);
+
+            leaf_heights.push(log2_strict(cur_leaf_len));
+
+            let num_tmp_digests = 2 * (cur_leaf_len - next_cap_len);
+
+            if cur_leaf_len == leaves_len {
+                cap = Vec::with_capacity(next_cap_len);
+                let tmp_cap_buf = capacity_up_to_mut(&mut cap, next_cap_len);
+                fill_digests_buf::<F, H>(
+                    &mut digests_buf[digests_buf_pos..(digests_buf_pos + num_tmp_digests)],
+                    tmp_cap_buf,
+                    &cur[..],
+                    next_cap_height,
+                );
+            } else {
+                //TODO: how to optimize it?
```

Comment:
So, I think this can possibly be directly optimized by switching from `&new_leaves[..]` to an iterator and ranges using `skip` and `take`. Not sure if that's the kind of thing you're asking for.

Reply:
I optimized it by using par_iter.

Reply:
I switched back to iter, since the map function is very simple.